markdown: Add support for `![alt text](url)` syntax for images.

N-Shar-ma opened the pull request:
markdown: Render inline images as small lightbox previews in place.

Images added using the `![alt text](url)` syntax are rendered in a fixed size, expandable via lightbox, exactly like the linked image previews, except inline.

For drafts which are only rendered by the frontend markdown processor, images are rendered as a regular link, for now.

markdown: Add support for `![alt text](url)` syntax for inline images.

The standard markdown syntax for images is now supported in Zulip. Unlike linked image previews (which is the only way images could be added so far), images added using the `![alt text](url)` syntax are truly inline, so are placed in the message exactly where the syntax is, not in a separate paragraph.

Fixes: #28912.

upload: Rename variable `filename_url` to `syntax_to_insert`.

This variable holds the markdown syntax for a file that's to be inserted into a message, hence the new name.

copy_and_paste: Fix bug where an unlabelled image could be pasted twice.

This commit removes the unnecessary condition for a linked image to have a label when checking whether to skip an image preview if its generating link has already been copied.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
Composing:
![image](https://github.com/zulip/zulip/assets/68962290/b8fed5ff-4282-45a2-9c84-41665fa5d04b)
Previewing:
![image](https://github.com/zulip/zulip/assets/68962290/bec82021-77ab-48e2-b192-2f3c0b945c90)
Drafted:
![image](https://github.com/zulip/zulip/assets/68962290/7e323258-b7de-413d-baa1-f038f49765d3)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on zerver/lib/markdown/__init__.py:
```diff
@@ -2133,6 +2139,43 @@ def handleMatch(  # type: ignore[override] # https://github.com/python/mypy/issu
+            div.set("class", "message_inline_image")
+            div.append(a)
+            return div, match_start, index
+        return None, None, None
```
When you write tests, let's make sure `InlineImageProcessor` is correctly processing these (should just work, but it is important for security).

N-Shar-ma replied:
I'm not sure I understand what you mean by "these" here? Do you just mean the general processing happening here?

timabbott replied:
I guess I want to make sure the `message_inline_image` objects generated by this syntax have `camo` processing on the URLs occur, which should just work.

N-Shar-ma replied:
Oh gotcha, the test added _does_ already check that!

timabbott commented:
When writing tests https://spec.commonmark.org/0.18/#images, may be a good source of things to try. Since we're using upstream code, we may not find it easy to match CommonMark on everything, but it should be cheap to at least know where we're diverging.

andersk commented on zerver/lib/markdown/__init__.py:
```diff
@@ -2133,6 +2139,43 @@ def handleMatch(  # type: ignore[override] # https://github.com/python/mypy/issu
+                # If the url is not a previewable image, gracefully fall back to a regular link
+                return self.link_processor.handleMatch(m, data)
+            # Else we wrap the image in an anchor tag and a div to enable lightbox preview.
+            alt = img.get("alt") or ""
```
```suggestion
            alt = img.get("alt", "")
```

andersk commented on zerver/lib/markdown/__init__.py:
```diff
@@ -2133,6 +2139,43 @@ def handleMatch(  # type: ignore[override] # https://github.com/python/mypy/issu
+            assert isinstance(img, Element)
+            src = img.get("src")
+            assert src is not None
+            src = sanitize_url(self.unescape(src.strip())) or ""
```
Do not use `src=""`. That causes the browser to load the _current page_ as an image, which could waste a lot of server time and bandwidth!

In general, blindly replacing an error with the empty string is usually not an appropriate way to handle an error.

andersk commented on zerver/lib/markdown/__init__.py:
```diff
@@ -2133,6 +2139,43 @@ def handleMatch(  # type: ignore[override] # https://github.com/python/mypy/issu
+            src = img.get("src")
+            assert src is not None
+            src = sanitize_url(self.unescape(src.strip())) or ""
+            if not any(src.endswith(ext) for ext in IMAGE_EXTENSIONS):
+                # If the url is not a previewable image, gracefully fall back to a regular link
+                return self.link_processor.handleMatch(m, data)
```
The `![…](…)` syntax indicates that the user knows this is an image. Not all images use an extension. We should give the user the image they asked for, not pretend we know “better”.

N-Shar-ma replied:
I'd observed the convert-unsupported-image-to-link behavior here in GitHub actually, but I now understand the merits of letting it be as is, thank you!

timabbott replied:
Can you give an example from GitHub of this detail?

N-Shar-ma replied:
@timabbott On writing something like `![image](image.png)` here in github, it renders as a plain link. Example: ![image](image.png)

andersk commented on zerver/lib/markdown/__init__.py:
```diff
@@ -2133,6 +2139,43 @@ def handleMatch(  # type: ignore[override] # https://github.com/python/mypy/issu
+            alt = img.get("alt") or ""
+            a = Element("a")
+            a.set("href", src)
+            a.set("aria-label", alt)
```
This is unnecessary since the `<img>` already has an `alt`. There’s no reason to force screen readers to read it twice.

N-Shar-ma replied:
The `aria-label` on the link is what is read for the tooltip, but yes, I'll deal with this on the frontend instead

andersk commented on zerver/lib/markdown/__init__.py:
```diff
@@ -2133,6 +2139,43 @@ def handleMatch(  # type: ignore[override] # https://github.com/python/mypy/issu
+            a = Element("a")
+            a.set("href", src)
+            a.set("aria-label", alt)
+            src = src.lstrip("/")
```
What?

N-Shar-ma replied:
This is needed for thumbnailing, had not added the condition and comments explaining this yet

andersk commented on zerver/lib/markdown/__init__.py:
```diff
@@ -2133,6 +2139,43 @@ def handleMatch(  # type: ignore[override] # https://github.com/python/mypy/issu
+            src = src.lstrip("/")
+            img.set("src", "/thumbnail?" + urlencode({"url": src, "size": "thumbnail"}))
+            img.set("data-src-fullsize", "/thumbnail?" + urlencode({"url": src, "size": "full"}))
+            a.append(img)
```
It’s the user’s responsibility to surround the image syntax with link syntax if they wanted it to be a link. And they might have already surrounded it with a link to somewhere else!

N-Shar-ma replied:
I'd taken this block of code from where we add images for previewable links in this same file. Now I understand why we'd want to avoid that. 

Thanks for pointing me to `rendered_markdown.update_elements`, I hadn't really gone through the frontend side of this change yet (hence the "WIP" description!)

andersk commented on zerver/lib/markdown/__init__.py:
```diff
@@ -2133,6 +2139,43 @@ def handleMatch(  # type: ignore[override] # https://github.com/python/mypy/issu
+            img.set("src", "/thumbnail?" + urlencode({"url": src, "size": "thumbnail"}))
+            img.set("data-src-fullsize", "/thumbnail?" + urlencode({"url": src, "size": "full"}))
+            a.append(img)
+            div = Element("div")
+            div.set("class", "message_inline_image")
+            div.append(a)
```
This is an inline processor, not a block processor. Wrapping the result in a `<div>` creates incorrectly nested HTML, since block elements auto-close the surrounding `<p>`.

And it’s not the behavior we want. Images are ![](https://upload.wikimedia.org/wikipedia/commons/thumb/b/bb/Kittyply_edit1.jpg/64px-Kittyply_edit1.jpg) inline in CommonMark.

N-Shar-ma replied:
Oh I see, since `.message_inline_image` are styled with `display: inline-block` on the frontend, I didn't realise they'd autoclose surrounding `<p>`s. Makes sense though, thanks for the insight!

N-Shar-ma commented:
I've addressed most of the feedback above, most importantly moved the html wrangling for presentation of inline images to the frontend, but I'm still working on refining things further. 

I'm yet to add backend tests, write a proper commit message, potentially split this across multiple commits, etc

N-Shar-ma commented:
Have added tests and cleaned up the code and commit structure, this is ready for another review!

N-Shar-ma commented:
> When writing tests https://spec.commonmark.org/0.18/#images, may be a good source of things to try. Since we're using upstream code, we may not find it easy to match CommonMark on everything, but it should be cheap to at least know where we're diverging.

The [reference style multiline syntax](https://spec.commonmark.org/0.18/#example-469) does not work, and any text inside `[]` is taken as is, so any inline syntax inside is copied to the image's alt attribute verbatim. These 2 are the main areas the implementation diverges from the spec. Other things like specifying the title attribute works as expected (see the added tests)

N-Shar-ma commented:
@zulipbot add "chat.zulip.org. review"

N-Shar-ma commented:
@zulipbot add "chat.zulip.org review"

timabbott commented:
I merged the first two commits as the series ending with 633d814df97662a39fe15d3ec22e0b4f2b3555d5.

timabbott commented:
So if you upload 3 images at once via dragging them or via selecting multiple files in the image picker, one gets two regression-looking outcomes:

- The images don't get put on the same line with each other, resulting in them stacking vertically rather than horizontally.
- The message controls seem to get displaced downwards in a message only containing such images.

![image](https://github.com/zulip/zulip/assets/2746074/49871030-3c02-4237-aa21-1fa5b61ea0fa)

N-Shar-ma commented:
> The images don't get put on the same line with each other, resulting in them stacking vertically rather than horizontally.

That is because we always put every upload (whether image or any other allowed format) in it's own line, so that's how they show up. What exact alternative behaviour are you proposing here @timabbott?

alya commented:
Hm, yeah, the image arrangement might be OK?

N-Shar-ma commented:
> The message controls seem to get displaced downwards in a message only containing such images.

I'm not sure why exactly it is needed currently, but removing the baseline alignment for `.messagebox-content` introduced in [this commit](https://github.com/zulip/zulip/pull/25997/commits/1ddaa23fa379dd9e287a343f0c2c82616fbf5959) seems to fix this. @karlstolley would appreciate your insight here as the author of this commit!

karlstolley commented:
That commit is necessary for the overall grid alignment in the message box. What I think needs to be applied here are the styles introduced in #26942 to the `&:has(> .message_inline_image:first-child)` selector to use the new paragraph-nested forms introduced here:

```css
    &:has(> .message_inline_image:first-child),
    &:has(> p > .message_inline_image:first-child) {
        align-self: center;
    }
```

It might be possible to tidy up and simplify that selector, but we'd have to watch very carefully for similar regressions that that might introduce.

Also, it does appear in testing that the fallback work with `@supports` already addresses this case, so no further adjustment is needed there.

timabbott commented:
For image arrangement, I think if you drag multiple images at once or select multiple in your browser in a single operation, it's probably more correct to put the images on a line together. Not sure if that's easy, but it was a major UX improvement when we made multiple linked images wrap on a single line (https://github.com/zulip/zulip/issues/20975), and this PR effectively undoes that feature from a practical standpoint.

N-Shar-ma commented:
@karlstolley Thanks a lot for linking to that PR, I faintly remembered it but couldn't find it. 

Since the problem was not for inline images that were 1st in their paragraph, but for any 1st paragraph containing an inline image, I've added the selector `&:has(> p:first-child > .message_inline_image)`. I tested extensively manually and it seems to work fine with a variety of message compositions.:

![image](https://github.com/zulip/zulip/assets/68962290/6d68b281-f4c2-41d1-ba73-7cd94f32e20d)

N-Shar-ma commented:
I did test for that too, just didn't include it in the screenshot.

![image](https://github.com/zulip/zulip/assets/68962290/91e79274-f352-4e21-a341-b95c4b671a01)

N-Shar-ma commented:
@timabbott I've added a prep commit that puts all uploads made in one go on the same line. This means for all kinds of uploads, not just images, each link text will be inserted inline together all in 1 block, while the previews will be as before (horizontally stacked in another paragraph). 

For the new `![alt text](url)` images, this means they too will stack horizontally when uploaded together

timabbott commented:
Hmm, this might be fairly ugly in the case that you upload a mixture of previewable elements and non-previewable elements.

N-Shar-ma commented:
@timabbott That's true :frowning_face: ... How about we only put them on 1 line if all the uploads are images (ie, anything we use the `![]()` syntax for)?

timabbott commented:
Yeah, that rule is probably a solid one.

N-Shar-ma commented:
@timabbott Made the change! Now only for uploads with multiple image-only type files, are the files all added in their own line, so they render as being stacked horizontally
